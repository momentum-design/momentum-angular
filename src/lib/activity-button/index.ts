export * from './activity-button.module';
export * from './activity-button.component';
