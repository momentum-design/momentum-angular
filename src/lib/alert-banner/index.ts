export * from './alert-banner.module';
export * from './alert-banner.component';
