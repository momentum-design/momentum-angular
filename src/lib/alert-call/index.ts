export * from './alert-call.module';
export * from './alert-call.component';
export * from './alert-call-config';
export * from './alert-call-container.component';
export * from './alert-call.service';
export * from './device-list-call.component';
