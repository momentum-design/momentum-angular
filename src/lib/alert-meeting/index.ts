export * from './alert-meeting.module';
export * from './alert-meeting.service';
export * from './alert-meeting-container.component';
export * from './alert-meeting-config';
export * from './alert-meeting.component';
