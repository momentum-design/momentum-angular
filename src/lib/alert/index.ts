export * from './alert.module';
export * from './alert.service';
export * from './alert-container.component';
export * from './alert-config';
export * from './alert.component';
