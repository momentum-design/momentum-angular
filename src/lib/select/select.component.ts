import {
  Component,
  Input,
  ViewChild,
  AfterContentChecked,
  ContentChildren,
  QueryList,
  AfterContentInit,
  OnInit,
  NgZone,
  Output,
  EventEmitter,
  HostListener,
  ElementRef,
  HostBinding
} from '@angular/core';
import {SelectionModel} from '@angular/cdk/collections';

import { uniqueId } from 'lodash';
import { ButtonComponent } from '../button';
import { ListItemComponent, OptionSelectionChange } from '../list-item';

import { takeUntil, switchMap, take, startWith } from 'rxjs/operators';
import { Subject, defer, Observable, merge } from 'rxjs';
import { ActiveDescendantKeyManager } from '@angular/cdk/a11y';
import {
  A,
  DOWN_ARROW,
  END,
  ENTER,
  HOME,
  SPACE,
  UP_ARROW,
  hasModifierKey,
} from '@angular/cdk/keycodes';
import { ControlValueAccessor } from '@angular/forms';

/** Change event object that is emitted when the select value has changed. */
export class SelectChange {
  constructor(
    public source: SelectComponent,
    public value: any
  ) { }
}

@Component({
  selector: 'cui-select',
  template: `
    <button cui-button
        cdkOverlayOrigin
        #trigger="cdkOverlayOrigin"
        [attr.name]='id'
        aria-label="select button"
        [id]='id'
        (click)='toggle()'
        [class]="buttonClasses"
        type="button"
        >
          <div class='cui-select__label' id="{{id}}__label">
              {{currentValue() || defaultValue}}
              <cui-icon name="arrow-down_16"></cui-icon>
          </div>
    </button>

    <ng-template
      cdkConnectedOverlay
      [cdkConnectedOverlayOrigin]="trigger"
      [cdkConnectedOverlayOpen]="overlayOpen"
      [cdkConnectedOverlayWidth]="this.anchorWidth"
      [cdkConnectedOverlayOffsetY]="6"
      [cdkConnectedOverlayPanelClass]="'cui-event-overlay__children'"
      [cdkConnectedOverlayHasBackdrop]="true"
      [cdkConnectedOverlayBackdropClass]="'cdk-overlay-transparent-backdrop'"
      (backdropClick)="close()"
      (detach)="close()">
      <cui-list
        #panel
        role='listbox'>
        <ng-content></ng-content>
      </cui-list>
    </ng-template>
  `,
  styles: [],
  host: {
    class: 'cui-input-group cui-select',
    '[class.disabled]': 'disabled'
  }
})
export class SelectComponent implements OnInit, AfterContentChecked, AfterContentInit, ControlValueAccessor {

  private _selectionModel: SelectionModel<ListItemComponent>;
  private readonly _destroy = new Subject<void>();

  /** Manages keyboard events for options in the panel. */
  _keyManager: ActiveDescendantKeyManager<ListItemComponent>;
  public overlayOpen = false;
  public anchorWidth = null;

  /** @prop Optional CSS button class name | '' */
  @Input() public buttonClass = '';
  /** @prop Set the default selected option | '' */
  @Input() public defaultValue: string = null;
  /** @prop Disable the Select Component | false */
  @HostBinding('attr.disabled') @Input() public disabled: boolean = false;
    /** @prop Set ID for Select Component | null */
  @Input() public id = uniqueId('cui-select-');
  /** @prop Optional prop to know if user is able to select multiple options | false */
  @Input() public isMulti = false;
  /** @option Callback function invoked by user selecting an interactive option within list | null */

  /** ngModel of Select Control */
  @Input()
  get ngModel(): any { return this._ngModelValue; }
  set ngModel(newValue: any) {
    if (newValue !== this._ngModelValue) {
      this.writeValue(newValue);
      this._ngModelValue = newValue;
    }
  }
  private _ngModelValue: any;

  @Output() select = new EventEmitter();

   /** Event emitted when the selected value has been changed by the user. */
   @Output() readonly selectionChange: EventEmitter<SelectChange> =
   new EventEmitter<SelectChange>();

  /** facilitates the two-way binding for the `ngModel` input. */
  @Output() readonly ngModelChange: EventEmitter<any> = new EventEmitter<any>();

  @ViewChild(ButtonComponent) originButton;

  /** Panel containing the select options. */
  @ViewChild('panel') panel: ElementRef;

  @ContentChildren(ListItemComponent) selectOptions: QueryList<ListItemComponent>;

  /** Combined stream of all of the child options' change events. */
  readonly optionSelectionChanges: Observable<OptionSelectionChange> = defer(() => {
    if (this.selectOptions) {
      return merge(...this.selectOptions.map(option => option.selectionChange));
    }

    return this._ngZone.onStable
      .asObservable()
      .pipe(take(1), switchMap(() => this.optionSelectionChanges));
  }) as Observable<OptionSelectionChange>;

  /** `View -> model callback called when value changes` */
  private _onChange: (value: any) => void = () => {};
  /** `View -> model callback called when select has been touched` */
  private _onTouched = () => {};
  /** Comparison function to specify which option is displayed. Defaults to object equality. */
  private _compareWith = (o1: any, o2: any) => o1 === o2;

  /** Handles keyboard events when the selected is open. */
  @HostListener('keydown', ['$event'])  _handleOpenKeydown = (event): void => {
    if (event.defaultPrevented || !this.overlayOpen) { return; }

    const keyCode = event.keyCode;
    const isArrowKey = keyCode === DOWN_ARROW || keyCode === UP_ARROW;
    const manager = this._keyManager;

    if (keyCode === HOME || keyCode === END) {
      event.preventDefault();
      keyCode === HOME ? manager.setFirstItemActive() : manager.setLastItemActive();
    } else if (isArrowKey && event.altKey) {
      // Close the select on ALT + arrow key to match the native <select>
      event.preventDefault();
      this.close();
    } else if ((keyCode === ENTER || keyCode === SPACE) && manager.activeItem &&
      !hasModifierKey(event)) {
      if (!this.isMulti) { this.close(); }
      event.preventDefault();
      manager.activeItem._selectViaInteraction();
    } else if (this.isMulti && keyCode === A && event.ctrlKey) {
      event.preventDefault();
      const hasDeselectedOptions = this.selectOptions.some(opt => !opt.disabled && !opt.selected);

      this.selectOptions.forEach(option => {
        if (!option.disabled) {
          hasDeselectedOptions ? option.select() : option.deselect();
        }
      });
    } else {
      const previouslyFocusedIndex = manager.activeItemIndex;

      manager.onKeydown(event);

      if (this.isMulti && isArrowKey && event.shiftKey && manager.activeItem &&
          manager.activeItemIndex !== previouslyFocusedIndex) {
        manager.activeItem._selectViaInteraction();
      }
    }
  }

  constructor(private _ngZone: NgZone) {
    this.optionSelectionChanges.subscribe(event => {
      this._onSelect(event.source, event.isUserInput);
    });
  }

  ngOnInit () {
    this._selectionModel = new SelectionModel<ListItemComponent>(this.isMulti);
  }

  ngAfterContentInit () {
    this._initKeyManager();

    this._selectionModel.changed.pipe(takeUntil(this._destroy)).subscribe(event => {
      event.added.forEach(option => option.select());
      event.removed.forEach(option => option.deselect());
    });

    this.selectOptions.changes.pipe(startWith(null), takeUntil(this._destroy)).subscribe(() => {
      this._resetOptions();
    });

    // set up children with props
    this.selectOptions.forEach((option) => {
      option.isMulti = this.isMulti;
      this._resetOptions();
    });
  }

  ngAfterContentChecked () {
    this._setAnchorWidth(this.originButton.el.nativeElement);
  }

  get buttonClasses (): string {
    return 'cui-button--input' +
    `${(this.buttonClass && ` ${this.buttonClass}`) || ''}` +
    `${(this.disabled && ` disabled`) || ''}`;
  }

  /** The currently selected option. */
  get selected(): ListItemComponent | ListItemComponent[] {
    return this.isMulti ? this._selectionModel.selected : this._selectionModel.selected[0];
  }

  /** Toggles the overlay panel open or closed. */
  toggle = (): void => {
    this.overlayOpen ? this.close() : this.open();
    this.focus();
  }

  open = (): void => {
    if (this.disabled || !this.selectOptions || !this.selectOptions.length || this.overlayOpen) {
      return;
    }
    this.overlayOpen = true;
  }

  close = (): void => {
    if (this.overlayOpen) {
      this.overlayOpen = false;
    }
  }

  /** Focuses the origin button of the select component. */
  focus = (): void => {
    this.originButton.el.nativeElement.focus();
  }

  currentValue = () => {
    if (!this._selectionModel) { return; }

    const selected = this._selectionModel.selected;
    if (!this.isMulti && selected.length) {
      return selected[0].label; }
    if (selected.length === 1) {
      return `${selected.length} Item Selected`;
    } else if (selected.length) {
      return `${selected.length} Items Selected`;
    }
  }

  /**
   * Sets the select's ngModelValue. Part of the ControlValueAccessor interface
   */
  writeValue = (value: any): void => {
    if (this.selectOptions) {
      this._setSelectionByValue(value);
    }
  }

  registerOnChange = (fn: (value: any) => void): void => {
    this._onChange = fn;
  }

  registerOnTouched = (fn: () => {}): void => {
    this._onTouched = fn;
  }

  setDisabledState ? = (isDisabled: boolean): void => {
    this.disabled = isDisabled;
  }

  /** Sets the selected option based on a value. */
  private _setSelectionByValue = (value: any | any[]): void => {
    if (this.isMulti && value) {
      if (!Array.isArray(value)) { throw Error('Value must be an array in multi-selection mode.'); }

      this._selectionModel.clear();
      value.forEach((currentValue: any) => this._selectValue(currentValue));
    } else {
      this._selectionModel.clear();
      const matchedOption = this._selectValue(value);

      if (matchedOption) {
        this._keyManager.setActiveItem(matchedOption);
      }
    }
  }

  private _selectValue = (value: any): ListItemComponent | undefined => {
    const matchedOption = this.selectOptions.find((option: ListItemComponent) => {
      try {
        return option.label != null && this._compareWith(option.label, value);
      } catch (error) {
        console.warn(error);
      }
      return false;
    });

    if (matchedOption) {
      this._selectionModel.select(matchedOption);
    }

    return matchedOption;
  }

  private _propagateChanges = (fallbackValue?: any): void => {
    let valueToEmit: any = null;

    if (this.isMulti) {
      valueToEmit = (this.selected as ListItemComponent[]).map(option => option.label);
    } else {
      valueToEmit = this.selected ? (this.selected as ListItemComponent).label : fallbackValue;
    }

    this._ngModelValue = valueToEmit;
    this.ngModelChange.emit(valueToEmit);
    this._onChange(valueToEmit);
    this.selectionChange.emit(new SelectChange(this, valueToEmit));
  }

  /** Sets up a key manager to listen to keyboard events on the overlay panel. */
  private _initKeyManager = (): void => {
    this._keyManager = new ActiveDescendantKeyManager<ListItemComponent>(this.selectOptions)
      .withWrap()
      .withTypeAhead();

    this._keyManager.tabOut.pipe(takeUntil(this._destroy)).subscribe(() => {
      // Restore focus to the trigger before closing
      this.focus();
      this.close();
    });

    this._keyManager.change.pipe(takeUntil(this._destroy)).subscribe(() => {
      if (!this.overlayOpen && !this.isMulti && this._keyManager.activeItem) {
        this._keyManager.activeItem._selectViaInteraction();
      }
    });
  }

  /** Drops current option subscriptions and IDs and resets from scratch. */
  private _resetOptions = (): void => {
    const changedOrDestroyed = merge(this.selectOptions.changes, this._destroy);

    this.optionSelectionChanges.pipe(takeUntil(changedOrDestroyed)).subscribe(event => {
      this._onSelect(event.source, event.isUserInput);

      if (event.isUserInput && !this.isMulti && this.overlayOpen) {
        this.close();
        this.focus();
      }
    });
  }

  /** Invoked when an option is clicked. */
  private _onSelect = (option: ListItemComponent, isUserInput: boolean): void => {
    const wasSelected = this._selectionModel.isSelected(option);
    if (option.label == null && !this.isMulti) {
      option.deselect();
      this._selectionModel.clear();
      this._propagateChanges(option.label);
    } else {
      option.selected ? this._selectionModel.select(option) : this._selectionModel.deselect(option);

      if (isUserInput) {
        this._keyManager.setActiveItem(option);
      }

      if (this.isMulti) {

        if (isUserInput) {
          // if clicked with mouse, focus on origin button
          this.focus();
        }
      }
    }
    if (wasSelected !== this._selectionModel.isSelected(option)) {
      this._propagateChanges();
    }
  }

  private _setAnchorWidth = (elementAnchor) => {
    const anchor = elementAnchor && elementAnchor.getBoundingClientRect();
    this.anchorWidth = anchor.width;
  }
}
