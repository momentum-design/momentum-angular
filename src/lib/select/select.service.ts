import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';

@Injectable()
export class SelectService {
  private selectionSource = new Subject();
  selectionSource$ = this.selectionSource.asObservable();

  onSelectionChange() {
    this.selectionSource.next();
  }
}
