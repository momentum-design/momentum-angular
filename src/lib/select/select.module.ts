import { NgModule, CUSTOM_ELEMENTS_SCHEMA, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SelectComponent } from './select.component';
import { ButtonModule } from '../button/button.module';
import { IconModule } from '../icon/icon.module';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { A11yModule } from '@angular/cdk/a11y';
import { ListModule } from '../list/list.module';
import { ListItemModule } from '../list-item/list-item.module';

@NgModule({
  imports: [
    CommonModule,
    ButtonModule,
    IconModule,
    PortalModule,
    ListModule,
    ListItemModule,
    OverlayModule,
    A11yModule,
  ],
  declarations: [
    SelectComponent,
  ],
  exports: [SelectComponent],
  schemas: [ CUSTOM_ELEMENTS_SCHEMA ]
})
export class SelectModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: SelectModule,
      providers: [],
    };
  }
}
