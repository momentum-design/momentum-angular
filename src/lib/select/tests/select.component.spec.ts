import { Component, QueryList, ContentChildren, ViewChild } from '@angular/core';
import { async, fakeAsync, ComponentFixture, TestBed, flush, inject } from '@angular/core/testing';
import { OverlayModule, OverlayContainer } from '@angular/cdk/overlay';
import { A11yModule } from '@angular/cdk/a11y';
import { ListItemComponent } from 'src/lib/list-item';
import {
  ButtonModule,
  IconModule,
  ListModule,
  ListItemSectionModule,
  CheckboxModule
} from 'src/lib/public_api';
import { SelectComponent } from '../select.component';
import { dispatchKeyboardEvent } from '../../../../test/utils';
import { DOWN_ARROW, ENTER } from '@angular/cdk/keycodes';

@Component({
  selector: 'select',
  template: `
        <md-select
          defaultValue="my default value"
          [ngClass]="'custom-ng-class'"
          [(ngModel)]="mySelectedItems">
            <div md-select-option class="custom-class-prop" label="first option"></div>
            <div md-select-option label="second option"></div>
        </md-select>
`
})
class TestSelectComponent {
  @ViewChild(SelectComponent) select: SelectComponent;
  @ContentChildren(ListItemComponent) selectOptions: QueryList<ListItemComponent>;

  mySelectedItems = '';
}

describe('SelectComponent', () => {
  let testComponent: TestSelectComponent;
  let selectComponent: SelectComponent;
  let fixture: ComponentFixture<TestSelectComponent>;
  let selectNativeElement: HTMLElement;
  let overlayContainer: OverlayContainer;
  let overlayContainerElement: HTMLElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        TestSelectComponent,
        SelectComponent,
        ListItemComponent
      ],
      imports: [
        ButtonModule,
        IconModule,
        ListModule,
        ListItemSectionModule,
        CheckboxModule,
        OverlayModule,
        A11yModule
      ]
    }).compileComponents();

    inject([OverlayContainer], (oc: OverlayContainer) => {
      overlayContainer = oc;
      overlayContainerElement = oc.getContainerElement();
    })();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestSelectComponent);
    testComponent = fixture.componentInstance;
    selectComponent = fixture.debugElement.children[0].componentInstance;
  });

  it('should create and match snapshot', () => {
    fixture.detectChanges();
    expect(testComponent).toBeTruthy();
    expect(selectComponent).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should render with default class name and ngClass', () => {
    fixture.detectChanges();

    selectNativeElement = fixture.nativeElement;
    const select = selectNativeElement.querySelector('md-select');
    expect(select).not.toBeNull();
    expect(select.className).toContain('md-input-group');
    expect(select.className).toContain('md-select');
    expect(select.className).toContain('custom-ng-class');
  });

  it('should render with defaultValue prop', () => {
    fixture.detectChanges();

    selectNativeElement = fixture.nativeElement;
    const select = selectNativeElement.querySelector('md-select');

    expect(select).not.toBeNull();
    expect(select.textContent.trim()).toEqual('my default value');
  });

  describe('button', () => {
    it('should render with default class name and buttonClass prop', () => {
      selectComponent.buttonClass = 'custom-button-class';
      fixture.detectChanges();

      selectNativeElement = fixture.nativeElement;
      const button = selectNativeElement.querySelector('button');
      expect(button).not.toBeNull();
      expect(button.className).toContain('md-button');
      expect(button.className).toContain('md-button--36');
      expect(button.className).toContain('custom-button-class');
    });

    it('should render with default randomly generated select id', () => {
      selectComponent.buttonClass = 'custom-button-class';
      fixture.detectChanges();

      selectNativeElement = fixture.nativeElement;
      const button = selectNativeElement.querySelector('button');
      expect(button).not.toBeNull();
      expect(button.id).toContain('md-select-');
    });
  });

  it('should open the overlay of options when select button is invoked', fakeAsync(() => {
    fixture.detectChanges();

    selectNativeElement = fixture.nativeElement;
    const button = selectNativeElement.querySelector('button');
    expect(button).not.toBeNull();

    // open overlay
    button.click();
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(true);
    expect(overlayContainerElement.textContent.trim()).toContain('first option');
    expect(overlayContainerElement.textContent.trim()).toContain('second option');
  }));

  it('should render two select options with default classname and class prop', fakeAsync(() => {
    fixture.detectChanges();

    selectNativeElement = fixture.nativeElement;
    const button = selectNativeElement.querySelector('button');
    expect(button).not.toBeNull();

    // open overlay
    button.click();
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(true);
    const selectOptions = selectComponent.selectOptions;
    expect(selectOptions).not.toBeNull();
    expect(selectOptions.length).toEqual(2);

    const firstOption = selectOptions.first;
    expect(firstOption.el.nativeElement.className).toContain('md-list-item');
    expect(firstOption.el.nativeElement.className).toContain('custom-class-prop');
    expect(fixture.componentInstance.mySelectedItems).toEqual('');
  }));

  it('should close the overlay of options when select button is invoked', fakeAsync(() => {
    fixture.detectChanges();

    selectNativeElement = fixture.nativeElement;
    const button = selectNativeElement.querySelector('button');
    expect(button).not.toBeNull();

    // open overlay
    button.click();
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(true);
    expect(overlayContainerElement.textContent.trim()).toContain('first option');
    expect(overlayContainerElement.textContent.trim()).toContain('second option');

    // close overlay
    button.click();
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(false);
    expect(overlayContainerElement.textContent.trim()).toEqual('');
  }));

  it('should close overlay when a click occurs outside the overlay', fakeAsync(() => {
    fixture.detectChanges();

    selectNativeElement = fixture.nativeElement;
    const button = selectNativeElement.querySelector('button');
    expect(button).not.toBeNull();

    button.click();
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(true);
    expect(overlayContainerElement.textContent.trim()).toContain('first option');
    const backdrop = overlayContainerElement.querySelector('.cdk-overlay-backdrop') as HTMLElement;

    backdrop.click();
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(false);
    expect(overlayContainerElement.textContent.trim()).toEqual('');
  }));

  it('should render correct text when first select option is clicked', fakeAsync(() => {
    fixture.detectChanges();

    selectNativeElement = fixture.nativeElement;
    const select = selectNativeElement.querySelector('md-select');
    const button = selectNativeElement.querySelector('button');
    expect(button).not.toBeNull();

    button.click();
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(true);
    const selectOptions = selectComponent.selectOptions;
    expect(selectOptions).not.toBeNull();

    const firstOption = selectOptions.first;
    expect(firstOption.selected).toBe(false);

    firstOption.handleClick(new Event('click'));
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(false);
    expect(firstOption.selected).toBe(true);
    expect(overlayContainerElement.textContent.trim()).toEqual('');
    expect(select.textContent.trim()).toEqual('first option');
    expect(fixture.componentInstance.mySelectedItems).toEqual('first option');
  }));

  it('should auto close overlay once an option is clicked', fakeAsync(() => {
    fixture.detectChanges();

    selectNativeElement = fixture.nativeElement;
    const button = selectNativeElement.querySelector('button');
    expect(button).not.toBeNull();

    button.click();
    fixture.detectChanges();
    flush();

    expect(overlayContainerElement.textContent.trim()).toContain('first option');
    const selectOptions = selectComponent.selectOptions;
    expect(selectOptions).not.toBeNull();

    const firstOption = selectOptions.first;
    expect(firstOption.selected).toBe(false);

    firstOption.handleClick(new Event('click'));
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(false);
    expect(firstOption.selected).toBe(true);
    expect(fixture.nativeElement.textContent.trim()).toEqual('first option');
    expect(overlayContainerElement.textContent.trim()).toEqual('');
  }));

  it('should render correctly when option is clicked and unclicked', fakeAsync(() => {
    fixture.detectChanges();

    selectNativeElement = fixture.nativeElement;
    const button = selectNativeElement.querySelector('button');
    expect(button).not.toBeNull();

    // open overlay
    button.click();
    fixture.detectChanges();
    flush();

    expect(selectComponent.overlayOpen).toBe(true);
    const selectOptions = selectComponent.selectOptions;
    expect(selectOptions).not.toBeNull();

    const firstOption = selectOptions.first;
    expect(firstOption.selected).toBe(false);
    expect(selectOptions.last.selected).toBe(false);
    let keyManager = fixture.componentInstance.select._keyManager;

    // select fist option
    firstOption.handleClick(new Event('click'));
    fixture.detectChanges();
    flush();

    // overlay is now closed
    expect(selectComponent.overlayOpen).toBe(false);
    expect(firstOption.selected).toBe(true);
    expect(selectOptions.last.selected).toBe(false);
    expect(fixture.nativeElement.textContent.trim()).toEqual('first option');
    expect(overlayContainerElement.textContent.trim()).toEqual('');
    expect(fixture.componentInstance.mySelectedItems).toEqual(selectOptions.first.label);

    // open overlay
    button.click();
    fixture.detectChanges();
    flush();

    // ensure first option is currently selected
    expect(selectOptions.first.selected).toBe(true);
    keyManager = fixture.componentInstance.select._keyManager;
    expect(keyManager.activeItem.id).toEqual(selectOptions.first.id);


    // invoked click event on first option to deselect
    firstOption.handleClick(new Event('click'));
    fixture.detectChanges();
    flush();

    expect(firstOption.selected).toBe(false);
    expect(selectOptions.last.selected).toBe(false);
    selectNativeElement = fixture.nativeElement;
    expect(selectNativeElement.textContent.trim()).toEqual('my default value');
    expect(fixture.componentInstance.mySelectedItems).toEqual(undefined);
  }));

  describe('accessibility', () => {
    it('keyboard enter should select the focused option', fakeAsync(() => {
      fixture.detectChanges();

      selectNativeElement = fixture.nativeElement;
      const button = selectNativeElement.querySelector('button');
      expect(button).not.toBeNull();

      // open overlay
      button.click();
      fixture.detectChanges();
      expect(selectComponent.overlayOpen).toBe(true);

      let selectOptions = selectComponent.selectOptions;

      selectNativeElement = fixture.nativeElement;
      const select = selectNativeElement.querySelector('md-select');

      dispatchKeyboardEvent(select, 'keydown', DOWN_ARROW);
      fixture.detectChanges();
      flush();

      // ensure first option is in focus
      selectOptions = selectComponent.selectOptions;
      expect(selectOptions.first.focus).toBe(true);

      dispatchKeyboardEvent(select, 'keydown', ENTER);
      fixture.detectChanges();
      flush();

      // after selection, overlay should be closed with updated select textContent
      expect(selectComponent.overlayOpen).toBe(false);
      expect(select.textContent.trim()).toEqual(selectOptions.first.label);

      // check ngModel value
      expect(fixture.componentInstance.mySelectedItems).toEqual('first option');

      // reopen overlay and confirm what is selected in the view
      button.click();
      expect(selectComponent.overlayOpen).toBe(true);

      selectOptions = selectComponent.selectOptions;
      const keyManager = fixture.componentInstance.select._keyManager;

      expect(selectOptions.first.selected).toBe(true);
      expect(keyManager.activeItem.id).toEqual(selectOptions.first.id);
    }));

    it('should resume focus from selected item after selecting via click', fakeAsync(() => {
      fixture.detectChanges();

      selectNativeElement = fixture.nativeElement;
      const button = selectNativeElement.querySelector('button');
      expect(button).not.toBeNull();

      // open overlay
      button.click();
      fixture.detectChanges();
      expect(selectComponent.overlayOpen).toBe(true);

      let selectOptions = selectComponent.selectOptions;
      selectNativeElement = fixture.nativeElement;
      const select = selectNativeElement.querySelector('md-select');

      dispatchKeyboardEvent(select, 'keydown', DOWN_ARROW);
      dispatchKeyboardEvent(select, 'keydown', DOWN_ARROW);
      fixture.detectChanges();
      flush();

      // ensure second option is in focus
      selectOptions = selectComponent.selectOptions;
      expect(selectOptions.last.focus).toBe(true);

      // select first option with a mouse click
      selectOptions = selectComponent.selectOptions;
      expect(selectOptions).not.toBeNull();
      const firstOption = selectOptions.first;
      expect(firstOption.selected).toBe(false);
      firstOption.handleClick(new Event('click'));

      fixture.detectChanges();
      flush();

      expect(selectComponent.overlayOpen).toBe(false);
      expect(select.textContent.trim()).toEqual(selectOptions.first.label);

      // check ngModel value
      expect(fixture.componentInstance.mySelectedItems).toEqual('first option');

      // reopen overlay and confirm what is selected in the view
      button.click();
      expect(selectComponent.overlayOpen).toBe(true);

      fixture.detectChanges();
      flush();

      selectOptions = selectComponent.selectOptions;
      const keyManager = fixture.componentInstance.select._keyManager;

      expect(selectOptions.first.selected).toBe(true);
      expect(keyManager.activeItem.id).toEqual(selectOptions.first.id);

      dispatchKeyboardEvent(select, 'keydown', DOWN_ARROW);
      fixture.detectChanges();
      flush();

      // ensure second option is in focus
      selectOptions = selectComponent.selectOptions;
      expect(selectOptions.last.focus).toBe(true);

      dispatchKeyboardEvent(select, 'keydown', ENTER);
      fixture.detectChanges();
      flush();

      // after selection, overlay should be closed with updated select textContent
      expect(selectComponent.overlayOpen).toBe(false);
      expect(select.textContent.trim()).toEqual(selectOptions.last.label);

      expect(fixture.componentInstance.mySelectedItems).toEqual('second option');
    }));
  });
});
