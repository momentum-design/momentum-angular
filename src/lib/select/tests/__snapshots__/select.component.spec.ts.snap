// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectComponent should create and match snapshot 1`] = `
<select
  mySelectedItems=""
  select={[Function SelectComponent]}
  selectOptions={[Function QueryList]}
>
  <md-select
    class="md-input-group md-select custom-ng-class"
    defaultvalue="my default value"
    disabled="false"
    ng-reflect-default-value="my default value"
    ng-reflect-ng-class="custom-ng-class"
    ng-reflect-ng-model=""
  >
    <button
      active="false"
      aria-label="select button"
      cdkoverlayorigin=""
      class="md-button md-button--36"
      href=""
      id="md-select-1"
      md-button=""
      name="md-select-1"
      ng-reflect-ng-class=""
      ng-reflect-type="button"
      tabindex="0"
      type="button"
    >
      
      <span
        class="md-button__children"
        ng-reflect-ng-style="[object Object]"
        style="opacity: 1;"
      >
        <div
          class="md-select__label"
          id="md-select-1__label"
        >
           my default value 
          <md-icon
            class="md-icon icon icon-arrow-down_16"
            name="arrow-down_16"
            ng-reflect-name="arrow-down_16"
          >
            
          </md-icon>
        </div>
      </span>
    </button>
    
  </md-select>
</select>
`;
